File: python/validation.py
"""JSON Schema validation for bridge request and response payloads."""

from jsonschema import Draft202012Validator

from schemas import (
    BRIDGE_SLUG,
    LOOKUP_BRIDGE_SLUG,
    LOOKUP_REQUEST_SCHEMA,
    LOOKUP_RESPONSE_SCHEMA,
    REQUEST_SCHEMA,
    RESPONSE_SCHEMA,
)

_REQUEST_VALIDATORS = {
    BRIDGE_SLUG: Draft202012Validator(REQUEST_SCHEMA),
    LOOKUP_BRIDGE_SLUG: Draft202012Validator(LOOKUP_REQUEST_SCHEMA),
}

_RESPONSE_VALIDATORS = {
    BRIDGE_SLUG: Draft202012Validator(RESPONSE_SCHEMA),
    LOOKUP_BRIDGE_SLUG: Draft202012Validator(LOOKUP_RESPONSE_SCHEMA),
}


def _field_name(error):
    """Derive the offending field name from a ``jsonschema`` ``ValidationError``.

    Falls back to parsing the message for ``required`` errors (which have
    no ``absolute_path``) and to ``"payload"`` when no field can be
    identified.
    """
    if error.absolute_path:
        return str(error.absolute_path[0])
    if error.validator == "required":
        return error.message.split("'")[1]
    return "payload"


def _format_errors(validator, payload):
    """Run ``validator`` over ``payload`` and return ``{name, message}`` dicts."""
    return [
        {"name": _field_name(error), "message": error.message}
        for error in validator.iter_errors(payload)
    ]


def validate_request(slug, payload):
    """Validate a request ``payload`` against the schema for ``slug``.

    Returns a list of ``{name, message}`` error dicts, empty on success.
    """
    return _format_errors(_REQUEST_VALIDATORS[slug], payload)


def validate_response(slug, payload):
    """Validate a response ``payload`` against the schema for ``slug``.

    Returns a list of ``{name, message}`` error dicts, empty on success.
    """
    return _format_errors(_RESPONSE_VALIDATORS[slug], payload)

File: node/validation.js
/**
 * @file JSON Schema validation for bridge request and response payloads.
 */

"use strict";

const Ajv = require("ajv/dist/2020");

const {
  BRIDGE_SLUG,
  LOOKUP_BRIDGE_SLUG,
  REQUEST_SCHEMA,
  RESPONSE_SCHEMA,
  LOOKUP_REQUEST_SCHEMA,
  LOOKUP_RESPONSE_SCHEMA,
} = require("./schemas");

const ajv = new Ajv({ allErrors: true, strict: false });

const REQUEST_VALIDATORS = {
  [BRIDGE_SLUG]: ajv.compile(REQUEST_SCHEMA),
  [LOOKUP_BRIDGE_SLUG]: ajv.compile(LOOKUP_REQUEST_SCHEMA),
};

const RESPONSE_VALIDATORS = {
  [BRIDGE_SLUG]: ajv.compile(RESPONSE_SCHEMA),
  [LOOKUP_BRIDGE_SLUG]: ajv.compile(LOOKUP_RESPONSE_SCHEMA),
};

/**
 * Derive the offending field name from an Ajv error object.
 *
 * Falls back to `params.missingProperty` for `required` errors (which
 * have no `instancePath`), to `params.additionalProperty` for
 * `additionalProperties` errors, and to `"payload"` when no field can
 * be identified.
 */
function fieldName(error) {
  if (error.instancePath) {
    return error.instancePath.replace(/^\//, "").split("/")[0];
  }
  if (error.keyword === "required") {
    return error.params.missingProperty;
  }
  if (error.keyword === "additionalProperties") {
    return error.params.additionalProperty;
  }
  return "payload";
}

function errorMessage(error) {
  if (error.keyword === "additionalProperties") {
    return `additional property '${error.params.additionalProperty}' is not allowed`;
  }
  return error.message;
}

/** Run `validator` over `payload` and return `{name, message}` objects. */
function formatErrors(validator, payload) {
  const valid = validator(payload);
  if (valid) return [];
  return (validator.errors || []).map((e) => ({
    name: fieldName(e),
    message: errorMessage(e),
  }));
}

/**
 * Validate a request `payload` against the schema for `slug`.
 *
 * Returns a list of `{name, message}` error objects, empty on success.
 */
function validateRequest(slug, payload) {
  return formatErrors(REQUEST_VALIDATORS[slug], payload);
}

/**
 * Validate a response `payload` against the schema for `slug`.
 *
 * Returns a list of `{name, message}` error objects, empty on success.
 */
function validateResponse(slug, payload) {
  return formatErrors(RESPONSE_VALIDATORS[slug], payload);
}

module.exports = { validateRequest, validateResponse };
